### PR TITLE
Mame: Update to 0.226

### PR DESCRIPTION
**Pull request by mascguy**
#### Description

Materials encompassing Mame 0.226 update.

Tracked by issue:
https://trac.macports.org/ticket/61482

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
macOS 10.12.6 16G2136
Xcode 9.2 9C40b

macOS 10.13.6 17G14019
Xcode 10.1 10B61

macOS 10.14.6 18G103
Xcode 11.3.1 11C505

macOS 10.15.6 19G2021
Xcode 12.0 12A7209

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by mascguy** [comment]
Folks, are the various CI build failures legitimate issues with this port? I reviewed the logs from the various jobs above, and the errors look like they could relate to the build system... with no details regarding this port.

No worries regardless, I just want to confirm whether this port is building successfully... and fix any issues, if necessary.

Thanks,
-Chris

File: emulators/mame/Portfile
Comment (ryandesign) [inline emulators/mame/Portfile:72]: You don't list checksums for files that are in the files directory of the port. MacPorts won't check them. MacPorts only checks checksums for files that are downloaded from master_sites or patch_sites.
```suggestion
                    size    196379874
```

File: emulators/mame/Portfile
Comment (ryandesign) [inline emulators/mame/Portfile:154]: Use the `configure.cxx_stdlib` variable, and only append this flag for compilers that understand it (clang). Append to the existing flags so as not to wipe out the flags MacPorts already sets. In fact, MacPorts already puts this flag in `configure.cxxflags` for you, so just remove this entirely?
```suggestion
```

File: emulators/mame/Portfile
Comment (ryandesign) [inline emulators/mame/Portfile:490]: Don't include literal newlines in the middle of paragraphs. Let MacPorts word wrap to the terminal width.
```suggestion
    If a blank screen is encountered, press ESC to exit, and then re-launch\
    with an alternative video option.
```

File: emulators/mame/Portfile
Comment (ryandesign) [inline emulators/mame/Portfile:460]: Destrooting the port failed:
```
Error: Failed to destroot mame: xinstall: Cannot stat: /opt/local/var/macports/distfiles/mame/mame-script-template-launcher.sh, No such file or directory
```
The file is not downloaded by MacPorts to distpath; you're providing it in filespath:
```suggestion
    xinstall ${filespath}/mame-script-template-launcher.sh \
```
(0755 is the default mode.)

**Review by ryandesign — CHANGES_REQUESTED** [review]

**Comment by ryandesign** [comment]
Thank you for doing the update but I will say you've vastly overcomplicated the port, or if not overcomplicated then at least made it a lot more complicated than it was before. I would anticipate that any other person thinking about updating the port in the future will change their mind once they open it and realize how complicated it is. If you want to take over as maintainer of the port and take responsibility for the complications and provide updates of the port in the future, that can be fine. But if you leave it as nomaintainer and don't come back to update the port in the future, I wouldn't be surprised if nobody ever updates the port again because of how complicated it has become.

**Comment by mascguy** [comment]
You’re absolutely right, it’s more complicated than it could be. That’s mainly due to my inexperience in the Portfile space. But as I become more comfortable/proficient in Portfile authoring, I’ll strive to keep things simpler.

And I greatly appreciate all of your support and feedback Ryan!

As for maintenance, I’d be happy to be the maintainer of this port.

Thanks again,
-Chris


On 2020-12-03-T, at 15:19, Ryan Schmidt <notifications@github.com> wrote:
Thank you for doing the update but I will say you've vastly overcomplicated the port, or if not overcomplicated then at least made it a lot more complicated than it was before. I would anticipate that any other person thinking about updating the port in the future will change their mind once they open it and realize how complicated it is. If you want to take over as maintainer of the port and take responsibility for the complications and provide updates of the port in the future, that can be fine. But if you leave it as nomaintainer and don't come back to update the port in the future, I wouldn't be surprised if nobody ever updates the port again because of how complicated it has become.

File: emulators/mame/Portfile
Comment (mascguy) [reply emulators/mame/Portfile:142]: Great feedback!

Currently verifying this locally, to be sure the resultant build flags are correct.

Comment (mascguy) [reply emulators/mame/Portfile:142]: Okay, now I remember why I replaced the standard flags: The compiler optimization level is included, which we don't want. That's because the Mame makefile controls that, via make variable 'OPTIMIZE'.

A few ports use the following to deal with that, which covers the four most common cases:
`configure.cxxflags-delete  -O1 -O2 -O3 -Os
`

Given that it's reasonably non-fragile, I'll combine the latter, with your earlier suggested change.

Comment (mascguy) [reply emulators/mame/Portfile]: Changes committed.

**Comment by mascguy** [comment]
Folks,

Barring the identification of any other critical issues, I'm finished committing changes for this release. Let me know if there's anything else I can do, relative to accepting these changes.

Thanks,
-Chris

File: emulators/mame/Portfile
Comment (mf2k) [inline emulators/mame/Portfile:57]: There is no need to name the download file. Please remove that (`mame-0226.tar.gz`). Otherwise this has to be updated everyone version update and that is not desirable.

File: emulators/mame/Portfile
Comment (mf2k) [inline emulators/mame/Portfile:62]: Please remove all of these extra comment spacing lines:

`#------------------------------------------------------------------------------
`

**Review by mf2k — CHANGES_REQUESTED** [review]

**Comment by kencu** [comment]
I'm building it now to see how it all goes.

File: emulators/mame/Portfile
Comment (kencu) [reply emulators/mame/Portfile:62]: are you strongly against those, Frank? He's put those in on purpose, obviously, as this is a complicated port that has a number of sections. And if he's willing to maintain mame properly, no skin off my nose if this makes it easier for him...

File: emulators/mame/Portfile
Comment (kencu) [reply emulators/mame/Portfile:57]: @macsguy - you only need to name your checksums if you have more than one.

**Comment by kencu** [comment]
having recently run up against this exact issue myself, at present, today, MacPorts is hoping that all ports will default to python38 rather than python39. I was asked to change one of mine 10 days ago, and did.

So that will happen.

File: emulators/mame/Portfile
Comment (mf2k) [reply emulators/mame/Portfile:62]: Not strongly no. This appears to be the most heavily commented Portfile I've seen in Macports. :) But that really is not a bad thing of course. Anyway, I'm fine with the comments themselves, but the padding seems unnecessary and is not in any other Portfile.

File: emulators/mame/Portfile
Comment (mascguy) [reply emulators/mame/Portfile:57]: Ah, this was a leftover from a few commits back, when I removed the checksums for all of the other artifacts.  :-)

File: emulators/mame/Portfile
Comment (mf2k) [reply emulators/mame/Portfile:62]: I'll defer to the maintainer on this.

Comment (mascguy) [reply emulators/mame/Portfile:62]: It's up to you guys. While I find the section separators helpful personally, I'm also a Portfile newbie. And I'd hate to see this be an example of Portfile Don'ts.  :-)

**Comment by mascguy** [comment]
> having recently run up against this exact issue myself, at present, today, MacPorts is hoping that all ports will default to python38 rather than python39. I was asked to change one of mine 10 days ago, and did.

No problem, the python39 default was an arbitrary choice on my part. So I'll change it to python38.

Thanks for the heads-up!

File: emulators/mame/Portfile
Comment (mascguy) [reply emulators/mame/Portfile:62]: Another thought: Since the Mame project is slated to release 0.227 within the next month or two -- and there are also some further refinements I'd like to make to this port -- I'm planning to release a new revision within the next month anyway.

So can a non-functional change like this, wait until then? Thoughts?

File: emulators/mame/Portfile
Comment (mascguy) [reply emulators/mame/Portfile]: Change committed, along with switching Python default from 3.9 to 3.8.

**Comment by mascguy** [comment]
> I'm building it now to see how it all goes.

I'm curious how long the build takes on your 2010 MacPro. On my 2008 model -- a dual quad-core 2.8 -- it takes 50-ish minutes.

**Comment by kencu** [comment]
> > I'm building it now to see how it all goes.
> 
> I'm curious how long the build takes on your 2010 MacPro. On my 2008 model -- a dual quad-core 2.8 -- it takes 50-ish minutes.

I started it when I made the note above, and checked now and it's done. So about an hour, indeed.

**Comment by kencu** [comment]
it all installed fine. I saw this:
```
$ mame
Unable to create history.db

Unable to create history.db

```
which I presume is from the subshell in a protected directory or something.

I didn't actually launch any games. Looks good, certainly more complete that what we have now. Actually runs. Man page works. 

We could probably go with this and then tweak later I guess. 

@mf2k ?

**Comment by mascguy** [comment]
> it all installed fine. I saw this:
> 
> ```
> $ mame
> Unable to create history.db
> 
> Unable to create history.db
> ```
> 
> which I presume is from the subshell in a protected directory or something.

Yes indeed, and it's one of the things I plan to correct in the next revision. At a minimum, I'll include more detail regarding Mame configuration, in the port notes. I'm also planning on including an example user configuration .INI file, which will eliminate such things. And anything else folks think makes sense. (Though I'll certainly reach out to you and others, before I make any concrete decisions.)

Great feedback, and thanks for verifying that it works locally!

**Comment by mascguy** [comment]
New ticket created, to formally track the usability improvements:

https://trac.macports.org/ticket/61743

File: emulators/mame/Portfile
Comment (mf2k) [reply emulators/mame/Portfile:61]: It's fine to leave them in this PR.

File: emulators/mame/Portfile
Comment (mf2k) [inline emulators/mame/Portfile:302]: To simplify things, please remove these old python version variants. Only the newer pythons are needed. Honestly I would remove them all and simply always use python 3.8.

**Review by mf2k — CHANGES_REQUESTED** [review]

**Comment by kencu** [comment]
chris - great work - is this your first PR. If so, outstanding. Frank -- I think we'll let Chris past-the-post here? he's really been pounding this into shape, and I was hoping to let him by for this and get it in, so we can perhaps move on with the tweaks on the next update. Always nice to see your work rewards,and I think he's earned it...

**Comment by kencu** [comment]
Now what's this commit history looking like ? we have 8 commits -- let me just see if i merge as is , or squash and merge.

**Comment by kencu** [comment]
I have to squash and merge, as the commit messages are not quite right for us.

File: emulators/mame/Portfile
Comment (mascguy) [reply emulators/mame/Portfile:302]: Added this to ticket 61739, covering internal cleanup:

https://trac.macports.org/ticket/61739

**Comment by mascguy** [comment]
Beautiful!

Folks, thanks for all of the support, feedback, and encouragement. You guys rock!

**Comment by mascguy** [comment]
Gents, I just checked for binary Mame packages, and thus far, they don't exist:

https://packages.macports.org/mame/?C=M;O=D

Is this as-expected, particularly given how long each Mame build takes?

(There's no rush. But I want to be sure that my portfile changes didn't somehow break/block the binary publish...)

**Comment by kencu** [comment]
because this port depends somewhere on openssl, it may not be distributable as a binary:
```
$ port rdeps mame | grep ssl
    openssl
```

This is a massive PITA for many ports. Various people have gone absolutely spare about it. Upstream has a new openssl release coming (3.0) "any day now" with a much more permissive license.

**Comment by kencu** [comment]
Every time this comes up, we wind up asking Josh whether or not it can be overridden. I don't understand myself the incantations needed to decide it one way or the other, but in the end, if it can be overridden, there is something you can add to the license to let it be distributable.

We should write this up as a FAQ as it comes up several times a month.

Somewhere there is a script that can analyze the deps and tell you what is wrong as well -- it doesn't seem to be in macports-contrib, so I forget just now where that script lives and who found it last time this came up.

**Comment by mascguy** [comment]
> Every time this comes up, we wind up asking Josh whether or not it can be overridden. I don't understand myself the incantations needed to decide it one way or the other, but in the end, if it can be overridden, there is something you can add to the license to let it be distributable.

@jmroot, would it be feasible to override this for the latest Mame port?

**Comment by jmroot** [comment]
The script used by the buildbot is https://github.com/macports/macports-infrastructure/tree/jobs/port_binary_distributable.tcl.

If all python is used for is generating the documentation, it's fine to do this:
```diff
diff --git a/emulators/mame/Portfile b/emulators/mame/Portfile
index 87b418076c6..81bde544bf7 100644
--- a/emulators/mame/Portfile
+++ b/emulators/mame/Portfile
@@ -228,6 +228,7 @@ proc mame_python_add_deps {} {
         "port:python${ver_major}${ver_minor}" \
         "port:py${ver_major}${ver_minor}-sphinx" \
         "port:py${ver_major}${ver_minor}-sphinxcontrib-svg2pdfconverter"
+    license_noconflict  python${ver_major}${ver_minor}
 
     return 0
 }
```

**Comment by mascguy** [comment]
Thanks guys!

New pull request created, covering this change:
https://github.com/macports/macports-ports/pull/9349

**Pull request merged**
